runtime/parachains/src/inclusion/mod.rs

```diff
@@ -793,6 +858,86 @@ impl<T: Config> Pallet<T> {
 			)
 	}
 
+	/// Check that all the upward messages sent by a candidate pass the acceptance criteria. Returns
+	/// false, if any of the messages doesn't pass.
+	pub(crate) fn check_upward_messages(
+		config: &HostConfiguration<T::BlockNumber>,
+		para: ParaId,
+		upward_messages: &[UpwardMessage],
+	) -> Result<(), UmpAcceptanceCheckErr> {
+		if upward_messages.len() as u32 > config.max_upward_message_num_per_candidate {
+			return Err(UmpAcceptanceCheckErr::MoreMessagesThanPermitted {
+				sent: upward_messages.len() as u32,
+				permitted: config.max_upward_message_num_per_candidate,
+			})
+		}
+
+		let fp = T::MessageQueue::footprint(para);
+		let (mut para_queue_count, mut para_queue_size) = (fp.count, fp.size);
+
+		for (idx, msg) in upward_messages.into_iter().enumerate() {
+			let msg_size = msg.len() as u32;
+			if msg_size > config.max_upward_message_size {
+				return Err(UmpAcceptanceCheckErr::MessageSize {
+					idx: idx as u32,
+					msg_size,
+					max_size: config.max_upward_message_size,
+				})
+			}
+			para_queue_count += 1;
+			para_queue_size += msg_size;
+		}
+
+		// make sure that the queue is not overfilled.
+		// we do it here only once since returning false invalidates the whole relay-chain block.
+		if para_queue_count > config.max_upward_queue_count {
+			return Err(UmpAcceptanceCheckErr::CapacityExceeded {
+				count: para_queue_count,
+				limit: config.max_upward_queue_count,
+			})
+		}
+		if para_queue_size > config.max_upward_queue_size {
+			return Err(UmpAcceptanceCheckErr::TotalSizeExceeded {
+				total_size: para_queue_size,
+				limit: config.max_upward_queue_size,
+			})
+		}
+
+		Ok(())
+	}
+
+	/// Enqueues `upward_messages` from a `para`'s accepted candidate block.
+	pub(crate) fn receive_upward_messages(
+		config: &HostConfiguration<T::BlockNumber>,
+		para: ParaId,
+		upward_messages: Vec<UpwardMessage>,
+	) -> Weight {
+		if !upward_messages.is_empty() {
+			let count = upward_messages.len() as u32;
+			Self::deposit_event(Event::UpwardMessagesReceived { from: para, count });
+		}
+		let messages = upward_messages.iter().filter_map(|d| BoundedSlice::try_from(&d[..]).ok());
+		T::MessageQueue::enqueue_messages(messages, para);
+
+		let Footprint { count, size } = T::MessageQueue::footprint(para);
+		// TODO: Do this at the end of the block, after any messages might have been
```

Comment:
does 'this' refer to the setting of well-known storage keys?

Reply:
it could be extracted into the `on_finalize` of the `initializer` pallet

Reply:
This is being done in the `on_queue_changed`.